util/src/io/mod.rs

```diff
@@ -0,0 +1,1089 @@
+                //
+                //   - Only the standard library gets to soundly "ignore" this,
+                //     based on its privileged knowledge of unstable rustc
+                //     internals;
+                g.buf.reserve(reservation_size(r));
+                let capacity = g.buf.capacity();
+                g.buf.set_len(capacity);
+                r.initializer().initialize(&mut g.buf[g.len..]);
+            }
+        }
+
+        match r.read(&mut g.buf[g.len..]) {
+            Ok(0) => {
+                ret = Ok(g.len - start_len);
+                break;
+            }
+            Ok(n) => g.len += n,
+            Err(ref e) if e.kind() == ErrorKind::Interrupted => {}
+            Err(e) => {
+                ret = Err(e);
+                break;
+            }
+        }
+    }
+
+    ret
+}
+
+#[cfg(feature = "alloc")]
+struct Guard<'a> {
+    buf: &'a mut Vec<u8>,
+    len: usize,
+}
+
+#[cfg(feature = "alloc")]
+impl Drop for Guard<'_> {
+    fn drop(&mut self) {
+        unsafe {
+            self.buf.set_len(self.len);
+        }
+    }
+}
+
+// A few methods below (read_to_string, read_line) will append data into a
+// `String` buffer, but we need to be pretty careful when doing this. The
+// implementation will just call `.as_mut_vec()` and then delegate to a
+// byte-oriented reading method, but we must ensure that when returning we never
+// leave `buf` in a state such that it contains invalid UTF-8 in its bounds.
+//
+// To this end, we use an RAII guard (to protect against panics) which updates
+// the length of the string when it is dropped. This guard initially truncates
+// the string to the prior length and only after we've validated that the
+// new contents are valid UTF-8 do we allow it to set a longer length.
+//
+// The unsafety in this function is twofold:
+//
+// 1. We're looking at the raw bytes of `buf`, so we take on the burden of UTF-8
+//    checks.
+// 2. We're passing a raw buffer to the function `f`, and it is expected that
+//    the function only *appends* bytes to the buffer. We'll get undefined
+//    behavior if existing bytes are overwritten to have non-UTF-8 data.
+#[cfg(feature = "alloc")]
+fn append_to_string<F>(buf: &mut String, f: F) -> Result<usize>
+where
+    F: FnOnce(&mut Vec<u8>) -> Result<usize>,
+{
+    unsafe {
+        let mut g = Guard {
+            len: buf.len(),
+            buf: buf.as_mut_vec(),
+        };
+        let ret = f(g.buf);
+        if str::from_utf8(&g.buf[g.len..]).is_err() {
+            ret.and_then(|_| {
+                Err(Error::new(
+                    ErrorKind::InvalidData,
+                    "stream did not contain valid UTF-8",
+                ))
+            })
+        } else {
+            g.len = g.buf.len();
+            ret
+        }
+    }
+}
+
+#[cfg(feature = "alloc")]
+fn read_until<R: BufRead + ?Sized>(r: &mut R, delim: u8, buf: &mut Vec<u8>) -> Result<usize> {
+    unimplemented!("eliza: figure out memchr!!!!")
```

Comment:
`core` actually contains a full-rust implementation of `memchr` (https://github.com/rust-lang/rust/blob/0fb43801368ae8b5931583f813071120bed55c35/src/libcore/slice/memchr.rs#L40-L83) to optimize the `[_]::contains` method for `u8` and `i8` (Thanks again @Manishearth!)

Unfortunately I don't think the full functionality is exposed anywhere stably, though we could copy the impl from libcore.